File: mns-core/src/main/java/cz/fb/manaus/core/model/TradedVolume.java
package cz.fb.manaus.core.model;

import com.google.common.collect.ImmutableList;
import com.google.common.primitives.Doubles;
import org.apache.commons.math3.stat.descriptive.moment.Mean;

import java.util.List;
import java.util.OptionalDouble;

import static java.util.stream.Collectors.toList;


public class TradedVolume {
    public static final TradedVolume EMPTY = new TradedVolume(List.of());

    private final List<Price> volume;

    public TradedVolume(Iterable<Price> volume) {
        this.volume = ImmutableList.copyOf(volume);
    }

    public static OptionalDouble getWeightedMean(List<Price> volume) {
        if (volume.isEmpty()) {
            return OptionalDouble.empty();
        } else {
            List<Double> prices = volume.stream().map(Price::getPrice).collect(toList());
            List<Double> amounts = volume.stream().map(Price::getAmount).collect(toList());
            return OptionalDouble.of(new Mean().evaluate(Doubles.toArray(prices), Doubles.toArray(amounts)));
        }
    }

    public OptionalDouble getWeightedMean() {
        return getWeightedMean(volume);
    }

    public List<Price> getVolume() {
        return volume;
    }
}

File: mns-core/src/main/java/cz/fb/manaus/core/model/TradedVolume.kt
package cz.fb.manaus.core.model

import com.fasterxml.jackson.annotation.JsonIgnore

data class TradedAmount(val price: Double, val amount: Double)

data class TradedVolume(val volume: List<TradedAmount>) {

    val weightedMean: Double?
        @JsonIgnore
        get() = getWeightedMean(volume)

    companion object {
        val EMPTY = TradedVolume(emptyList())
    }
}


fun getWeightedMean(volume: List<TradedAmount>): Double? {
    return getWeightedMean(volume, TradedAmount::price, TradedAmount::amount)
}
